nucliadb_node/nucliadb_node/listeners/gc_scheduler.py

```diff
@@ -0,0 +1,147 @@
+#
+# You should have received a copy of the GNU Affero General Public License
+# along with this program. If not, see <http://www.gnu.org/licenses/>.
+
+import asyncio
+from functools import partial
+from typing import Optional
+
+from nucliadb_node import logger, signals
+from nucliadb_node.settings import settings
+from nucliadb_node.signals import SuccessfulIndexingPayload
+from nucliadb_node.writer import Writer
+from nucliadb_telemetry import metrics
+
+gc_observer = metrics.Observer(
+    "gc_processor",
+    buckets=[
+        0.025,
+        0.05,
+        0.1,
+        1.0,
+        5.0,
+        10.0,
+        30.0,
+        60.0,
+        120.0,
+        300.0,
+        float("inf"),
+    ],
+)
+
+
+class ShardManager:
+    schedule_delay_seconds = 30.0
+
+    def __init__(self, shard_id: str, writer: Writer, gc_lock: asyncio.Semaphore):
+        self.lock = asyncio.Lock()  # write lock for this shard
+        self.target_gc_resources = settings.max_resources_before_gc
+
+        self._shard_id = shard_id
+        self._writer = writer
+
+        self._gc_lock = gc_lock  # global lock so we only gc one shard at a time
+        self._change_count = 0
+        self._gc_schedule_timer: Optional[asyncio.TimerHandle] = None
+        self._gc_task: Optional[asyncio.Task] = None
+
+    def shard_changed_event(self, delay: Optional[float] = None) -> None:
+        """
+        Signal shard has changed and should be garbage collected at some point
+        """
+        if self._gc_task is not None and not self._gc_task.done():
+            # already running or scheduled
+            return
+
+        self._change_count += 1
+        if (
+            self._gc_schedule_timer is not None
+            and not self._gc_schedule_timer.cancelled()
+        ):
+            self._gc_schedule_timer.cancel()
+
+        if self._change_count >= self.target_gc_resources:
+            # need to force running it now
+            self._schedule_gc()
+        else:
+            # run it soon
+            if delay is None:
+                delay = self.schedule_delay_seconds
+            self._gc_schedule_timer = asyncio.get_event_loop().call_later(
+                delay, self._schedule_gc
+            )
+
+    def _schedule_gc(self) -> None:
+        self._gc_task = asyncio.create_task(self.gc())
+
+    async def gc(self):
+        async with self._gc_lock, self.lock:
+            logger.info("Running garbage collection", extra={"shard": self._shard_id})
+            self._change_count = 0
+            try:
+                with gc_observer():
+                    # NOTE: garbage collector may not run if the shard is busy. We currently don't do anything to retry.
+                    await self._writer.garbage_collector(self._shard_id)
+                    logger.info(
+                        "Garbage collection finished", extra={"shard": self._shard_id}
+                    )
+            except Exception:
+                logger.exception(
```

Comment:
do we want to send a sentry event with errors.capture_exception() in this case?

Reply:
Added a `capture_exception`